### PR TITLE
Fix Reactive Route stream serialization without relying on the Multi sub-type

### LINKED ISSUE #21044: Reactive routes incorrect json serialization
### Describe the bug

Reactive routes resources returning invalid json when using `@RolesAllowed` annotation. Without the annotation its working fine.

Its working fine using Quarkus 2.3 version

### Expected behavior

API should return below JSON response

```javascript
[
	{
		"message": "Hello!"
	},
	{
		"message": "Hola!"
	},
	{
		"message": "Bonjour!"
	},
	{
		"message": "Ciao!"
	}
]
```

### Actual behavior

Instead its returning

```javascript
{"message":"Hello!"}{"message":"Hola!"}{"message":"Bonjour!"}{"message":"Ciao!"}
```

### How to Reproduce?

Reproducer: https://github.com/MM87037/reactive-routes

Steps:

1. Run `mvn clean verify` to see the failure

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

OpenJDK Runtime Environment Temurin-17+35 (build 17+35)
OpenJDK 64-Bit Server VM Temurin-17+35 (build 17+35, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.4.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by cescoffier**
Provide another way to generate streams from Reactive Routes that do not require a specific Multi sub-type.

Sub-typing Multi does not work when the route is protected using @RolesAllowed as the security interceptor changes the returned Multi actual type.

The previous approach is still supported but is deprecated. The documentation mentions the new approach and explains why the previous approach should be avoided.

CC @michalszynkiewicz 

Fix #21044

### CODE REVIEW

**Comment by mkouba** [comment]
This a breaking change, right? Because previously, the `Route#produces()` value was only used for content based routing.

**Comment by cescoffier** [comment]
Yes, it is breaking if you were configuring ‘produces.’ However, in practice, it should not happen too much (still needs to be mentioned.

File: extensions/reactive-routes/runtime/src/main/java/io/quarkus/vertx/web/ReactiveRoutes.java [hunk 1]
```diff
@@ -12,6 +12,74 @@
...
+     *             new Person("batman", 2),
+     *             new Person("spiderman", 3));
+     * }
+     * }
+     * </pre>
+     *
+     * Note that the array is streamed object per object.
+     * Each object is written individually in the response, until the last one.
+     */
+    public static final String APPLICATION_JSON_CONTENT_TYPE = "application/json";
```
Comment (mkouba) [inline extensions/reactive-routes/runtime/src/main/java/io/quarkus/vertx/web/ReactiveRoutes.java:33]: Doesn't vert.x web define some ContentType constants? Anyway, I think that we could use a shorter name such as `APPLICATION_JSON`.

File: docs/src/main/asciidoc/reactive-routes.adoc [hunk 2]
```diff
@@ -399,18 +399,16 @@ The previous snippet produces:
...
 The response is written item by item to the client.
-The `content-type` is set to `application/json` if not set already.
-
-To use this feature, you need to wrap the returned `Multi` using `io.quarkus.vertx.web.ReactiveRoutes.asJsonArray`:
+To do that set the `produces` attribute to `"application/json"` (or `ReactiveRoutes.APPLICATION_JSON_CONTENT_TYPE`).
 
 [source, java]
 ----
-@Route(path = "/people")
+@Route(path = "/people", produces = ReactiveRoutes.APPLICATION_JSON_CONTENT_TYPE)
```
Comment (mkouba) [inline docs/src/main/asciidoc/reactive-routes.adoc:406]: I think that we should also mention `ReactiveRoutes.APPLICATION_JSON_CONTENT_TYPE` and friends in the `Route#produces` javadoc.

**Review by mkouba — COMMENTED** [review]
I've added few minor comments..

File: extensions/reactive-routes/runtime/src/main/java/io/quarkus/vertx/web/ReactiveRoutes.java (hunk 1 above)
Comment (cescoffier) [reply extensions/reactive-routes/runtime/src/main/java/io/quarkus/vertx/web/ReactiveRoutes.java:33]: No there are no constants we can reuse, but I've reduced the name length.

File: extensions/reactive-routes/deployment/src/main/java/io/quarkus/vertx/web/deployment/ReactiveRoutesProcessor.java [hunk 3]
```diff
@@ -775,33 +779,48 @@ void implementInvoke(HandlerDescriptor descriptor, BeanInfo bean, MethodInfo met
...
+                } else if (contentType.toLowerCase().startsWith(APPLICATION_JSON_CONTENT_TYPE)) {
+                    handleJsonArrayMulti(descriptor, invoke, routingContext, res);
+                } else if (contentType.toLowerCase().startsWith(ND_JSON_CONTENT_TYPE)
+                        || contentType.toLowerCase().startsWith(JSON_STREAM_CONTENT_TYPE)) {
+                    handleNdjsonMulti(descriptor, invoke, routingContext, res);
+                } else {
+                    handleRegularMulti(descriptor, invoke, routingContext, res);
+                }
+            } else {
+                // No content type, use the Multi Type - this approach is not reliable and should be avoided.
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply extensions/reactive-routes/deployment/src/main/java/io/quarkus/vertx/web/deployment/ReactiveRoutesProcessor.java]: I rephrase it to say that it does not work when Quarkus security is used as it wraps the returned Mutli.

File: docs/src/main/asciidoc/reactive-routes.adoc (hunk 2 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/reactive-routes.adoc]: Added, with explanations.

**Review by mkouba — APPROVED** [review]

**Comment by mkouba** [comment]
@cescoffier Pls add a note to the migration guide.

**Comment by cescoffier** [comment]
@mkouba Done: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.6#reactive-routes-produces-changes
